llama-index-core/llama_index/core/evaluation/retrieval/omni_modal.py

```diff
@@ -0,0 +1,285 @@
+import asyncio
+import json
+from dataclasses import asdict, dataclass
+from typing import Any, Callable, Collection, Dict, Generic, List, Optional, Tuple
+
+from llama_index.core.async_utils import asyncio_run
+from llama_index.core.bridge.pydantic import BaseModel, Field
+from llama_index.core.embeddings.omni_modal_base import (
+    KD,
+    KQ,
+)
+from llama_index.core.evaluation.retrieval.metrics import resolve_metrics
+from llama_index.core.evaluation.retrieval.metrics_base import (
+    BaseRetrievalMetric,
+    RetrievalMetricResult,
+)
+from llama_index.core.indices.omni_modal import OmniModalVectorIndexRetriever
+from llama_index.core.postprocessor.types import BaseNodePostprocessor
+from llama_index.core.schema import BaseNode, QueryBundle
+
+
+class OmniModalEmbeddingQAFinetuneDataset(BaseModel):
+    """Omni-Modal Embedding QA Finetuning Dataset.
+
+    Args:
+        queries (Dict[str, QueryBundle]): Dict id -> query.
+        corpus (Dict[str, BaseNode]): Dict id -> string.
+        relevant_docs (Dict[str, List[str]]): Dict query id -> list of doc ids.
+
+    """
+
+    queries: Dict[str, QueryBundle]  # dict id -> query
+    corpus: Dict[str, BaseNode]  # dict id -> string
+    relevant_docs: Dict[str, List[str]]  # query id -> list of doc ids
+
+    @property
+    def query_docid_pairs(self) -> List[Tuple[QueryBundle, List[str]]]:
+        """Get query, relevant doc ids."""
+        return [
+            (query, self.relevant_docs[query_id])
+            for query_id, query in self.queries.items()
+        ]
+
+    def save_json(self, path: str) -> None:
+        """Save json."""
+        data = {
+            "queries": {k: asdict(v) for k, v in self.queries.items()},
+            "corpus": {k: v.to_dict() for k, v in self.corpus.items()},
+            "relevant_docs": self.relevant_docs,
+        }
+
+        with open(path, "w") as f:
+            json.dump(data, f, indent=4)
+
+    @classmethod
+    def from_json(
+        cls,
+        path: str,
+        *,
+        query_loader: Callable[[Dict[str, Any]], QueryBundle] = QueryBundle.from_dict,
```

Comment:
How should we serialization/deserialize multi-modal queries and documents (and any class that contains them, like this `OmniModalEmbeddingQAFinetuneDataset`) without requiring the user to pass in the concrete classes to load?